python/lsst/ts/ess/earthquake/q330_connector.py

```diff
@@ -178,28 +348,66 @@ async def connect(self) -> None:
-        self.log.debug("Q330 state is not None.")
-        # Now wait until the sensor is ready to start.
-        while not (
-            self.q330_state.state_type == TStateType.ST_STATE.value
-            and self.q330_state.info == TLibState.LIBSTATE_RUNWAIT.value
-        ):
-            await asyncio.sleep(0.1)
-        self.log.debug("Ready to send telemetry.")
-        # Ready, so start the telemetry stream.
-        self.libq330.q330_change_state(TLibState.LIBSTATE_RUN.value)
-        self.log.debug("Telemetry stream started.")
+
+        await self.stop_telemetry_thread()
+        self.log.debug("Starting telemetry thread.")
+        self.telemetry_thread = threading.Thread(target=self._telemetry_worker)
+        self.telemetry_thread.start()
+        self.log.debug("Done starting telemetry thread.")
+
+    def _telemetry_worker(self) -> None:
+        telemetry_loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(telemetry_loop)
+        self.log.debug("Starting telemetry loop.")
+        asyncio.run(self._process_telemetry())
+
+    async def _process_telemetry(self) -> None:
+        """Process pending telemetry.
+
+        Check the queue for pending telemetry and write it if there is any.
+        """
+        if not self.telemetry_event:
+            raise RuntimeError("No telemetry event.")
+        while not self.telemetry_event.is_set():
+            self.log.debug(
+                f"Length of telemetry queue = {self.telemetry_queue.qsize()}"
+            )
+            try:
+                # Just poll until the queue is empty.
+                while True:
+                    cdh = self.telemetry_queue.get(False)
+                    topic = self._get_topic_for_name(cdh.topic_name)
+                    self.log.debug(
+                        f"Writing [{cdh.topic_name=}, {cdh.timestamp=}, "
+                        f"{cdh.e=}, {cdh.n=}, {cdh.z=}] telemetry."
+                    )
+                    await topic.set_write(
+                        timestamp=cdh.timestamp, e=cdh.e, n=cdh.n, z=cdh.z
+                    )
+            except queue.Empty:
+                # Ignore an empty queue.
+                pass
+            await asyncio.sleep(TELEMETRY_WAIT)
 
     async def disconnect(self) -> None:
         """Disconnect from the earthquake sensor.
 
         Disconnecting is performed by instructing the lib330 library to destroy
         the conneciton context.
         """
+        await self.stop_telemetry_thread()
+
         self.libq330.q330_change_state(TLibState.LIBSTATE_IDLE.value)
         await asyncio.sleep(1.0)
         self.libq330.q330_change_state(TLibState.LIBSTATE_TERM.value)
         await asyncio.sleep(1.0)
         self.libq330.q330_destroy_context()
+
+    async def stop_telemetry_thread(self) -> None:
+        self.telemetry_event.clear()
+        if self.telemetry_thread:
+            self.log.debug("Stopping telemetry thread.")
+            self.telemetry_event.set()
```

Comment:
Python doesn't have an API to kill a thread. It would require to use a subprocess instead. The preferred mechanism for stopping a thread is to set a flag and check if the flag is set. I modified both the loop in the thread accordingly to your suggestions above and also to check for the Event in a different way. FWIW I have followed this blog post:
https://www.pythontutorial.net/python-concurrency/python-stop-thread/
Having said that, I introduced a timeout of 5 seconds for the thread to stop.